upnp.go

```diff
@@ -113,18 +115,27 @@ func (d *IGD) getInternalIP() (string, error) {
 }
 
 // Discover scans the local network for routers and returns the first
-// UPnP-enabled router it encounters.
+// UPnP-enabled router it encounters.  It will try up to 3 times to find a
+// router, sleeping a random duration between each attempt.  This is to
+// mitigate a race condition with many callers attempting to discover
+// simultaniously.
 //
 // TODO: if more than one client is found, only return those on the same
 // subnet as the user?
 func Discover() (*IGD, error) {
-	pppclients, _, _ := internetgateway1.NewWANPPPConnection1Clients()
-	if len(pppclients) > 0 {
-		return &IGD{pppclients[0]}, nil
-	}
-	ipclients, _, _ := internetgateway1.NewWANIPConnection1Clients()
-	if len(ipclients) > 0 {
-		return &IGD{ipclients[0]}, nil
+	maxTries := 3
+	sleepTime, _ := crypto.RandIntn(5)
```

Comment:
needs to be milliseconds or nanoseconds, full seconds is not going to give you as great of a backoff distribution